src/ekat/logging/ekat_logger.hpp

```diff
@@ -0,0 +1,158 @@
+
+  Logs have a priority logging level.  Messages with priorities below this level will
+  be ignored.
+
+  Logging levels (increasing priority):
+  - off
+  - trace
+  - debug
+  - info
+  - warn
+  - err
+  - critical
+
+  The default priority for the spdlog library is "info."  For EKAT, we've set the
+  default to "debug."  Both are easy to change.
+
+  The LogMpiPolicy can restrict console output to only rank 0, and it will append
+  rank ids to log files and log names.
+
+  The LogFilePolicy determines whether or not the log will also write output to a file.
+  Files and console can have different logging levels.  For example, the
+  log level for the LogBasicFile policy could be "debug," while the level of the console
+  logger could be set to "info".  In this case, no debug messages will show up on the
+  console, but they will be in the log files.
+
+  In principle, separate modules, classes, etc. could have their own logger.
+  These loggers can share output files, too; see the spdlog wiki
+(https://github.com/gabime/spdlog/wiki/2.-Creating-loggers#creating-multiple-file-loggers-with-the-same-output-file) for an example.
+
+  The ekat::Logger class is a subclass of spdlog::logger class, to simplify the many
+  features of that library to the ones EKAT is most likely to use, while retaining all
+  of its functionality.
+*/
+template <typename LogFilePolicy = LogNoFile,    // see ekat_log_file.hpp
+          typename LogMpiPolicy = LogIgnoreRank> // see ekat_log_mpi.hpp
+class Logger : public spdlog::logger {
+  public:
+    Logger() = delete;
+
+    Logger(const std::string& log_name, const std::string& level_str="debug") :
+     spdlog::logger(LogMpiPolicy::name_append_rank(log_name))
+  {
+
+    const auto clevel = Log::level::from_str(level_str);
+    auto csink = std::make_shared<spdlog::sinks::stdout_color_sink_mt>();
+    csink->set_level(clevel);
+
+    std::string fname = LogMpiPolicy::name_append_rank(log_name) + "_logfile.txt";
+    auto fsink = LogFilePolicy::get_file_sink(fname);
+
+    this->sinks().push_back(csink);
+    this->sinks().push_back(fsink);
+
+    LogMpiPolicy::update_sinks(this->sinks());
+
+    this->set_level(clevel);
+  }
+
+  spdlog::sink_ptr get_console_sink() {
+    return (this->sinks().size() > 1 ? this->sinks()[0] : spdlog::sink_ptr(nullptr));
+  }
+
+  spdlog::sink_ptr get_file_sink() {
+    return (this->sinks().size() > 1 ? this->sinks()[1] : this->sinks()[0]);
+  }
+
+  template <typename FP = LogFilePolicy>
+  typename std::enable_if<FP::has_filename, std::string>::type
+  get_logfile_name() {
+    std::string result;
+    auto fsink = this->get_file_sink();
+    spdlog::sinks::basic_file_sink_mt* bptr =
+      dynamic_cast<spdlog::sinks::basic_file_sink_mt*>(fsink.get());
+    spdlog::sinks::rotating_file_sink_mt* rptr =
+      dynamic_cast<spdlog::sinks::rotating_file_sink_mt*>(fsink.get());
+    if (bptr) {
+      result = bptr->filename();
+    }
+    else if (rptr) {
```

Comment:
You might do the second cast in here, and maybe throw if the second cast fails (since we expect a file sink to be either basic or rotatin, right?).

Reply:
Yeah, it would be nice if null_sink returned an empty string for its filename method, same for console sink.

Reply:
> You might do the second cast in here, and maybe throw if the second cast fails (since we expect a file sink to be either basic or rotatin, right?)

The file sink can be `spdlog::null_sink_mt`, which does not have a `filename()` method, so it would be incorrect to throw.  Maybe it's worth asking the `spdlog` folks to add a `filename()` method to `null_sink_mt` that just returns an empty string.

Reply:
Ah, good point.